src/utils/index.ts

```diff
@@ -152,24 +161,34 @@ export function formatData(data: Data, args: Inputs): Output {
     // Pass any required query params with user values for relevant scripts
     scripts: data.scripts
       ? data.scripts.map((script) => {
+          const paramKeys = script.params;
+          const paramInputs = filterArgs(args, paramKeys);
+          const optionalParamKeys = script.optionalParams
+            ? Object.keys(script.optionalParams)
+            : [];
+
+          // if we're receiving undefined or null as an ar, we check the optionalParams
+          const newArgs: any = { ...script.optionalParams };
+          Object.keys(args).forEach((key) => {
+            if (args[key]) newArgs[key] = args[key];
+          });
+          const optionalParamInputs = filterArgs(newArgs, optionalParamKeys);
```

Comment:
Okay, it looks like here you're merging the optional param defaults with the actual optional param values provided by the caller. That sort of satisfies my comment above, but I think that's still error-prone IMO, as it means this only works out of the box if you call `formatData()`, but not if you call `formatUrl()` directly.

So I think it would still be a safer implementation to pass the object with optional param keys and their defaults to `formatUrl` and handle the "merging" in there, so that the function can be used intuitively.